FWCore/Concurrency/interface/WaitingThreadPool.h

```diff
@@ -0,0 +1,106 @@
+#ifndef FWCore_Concurrency_WaitingThreadPool_h
+#define FWCore_Concurrency_WaitingThreadPool_h
+
+#include "FWCore/Utilities/interface/ConvertException.h"
+#include "FWCore/Utilities/interface/ReusableObjectHolder.h"
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace edm {
+  namespace impl {
+    class WaitingThread {
+    public:
+      WaitingThread();
+      ~WaitingThread() noexcept;
+
+      WaitingThread(WaitingThread const&) = delete;
+      WaitingThread& operator=(WaitingThread&&) = delete;
+      WaitingThread(WaitingThread&&) = delete;
+      WaitingThread& operator=(WaitingThread const&) = delete;
+
+      template <typename F, typename G>
+      void run(WaitingTaskWithArenaHolder holder,
+               F&& func,
+               G&& errorContextFunc,
+               std::shared_ptr<WaitingThread> thisPtr) {
```

Comment:
Should this be `noexcept`?

Reply:
Depends if we want possible system errors from locking the mutex to be reported via exceptions to be caught by the caller, or via `std::terminate()`.

Reply:
Let's leave as it is.